kpi/serializers/v2/asset_file.py

```diff
@@ -40,8 +46,262 @@ class Meta:
+        return ret
+
+    def validate(self, attr):
+        self.__file_type = attr['file_type']  # noqa
+
+        metadata = self._get_metadata(attr.get('metadata'))
+        validated_field = self._validate_media_content_method(attr, metadata)
+        # Call the validator related to `validated_field`, either:
+        # - `self._validate_content()`
+        # - `self._validate_base64Encoded()`
+        # - `self._validate_redirect_url()`
+        validator = getattr(self, f'_validate_{validated_field}')
+        validator(attr, metadata)
+
+        # Common validators
+        filename = metadata['filename']
+        self.__validate_mime_type(filename, validated_field)
+        self.__validate_extension(filename, validated_field)
+        self._validate_duplicate(filename, validated_field)
+
+        # Remove `'base64Encoded'` from attributes passed to the model
+        attr.pop('base64Encoded', None)
+
+        return attr
+
+    def _get_metadata(self, metadata: Union[str, dict]) -> dict:
+        """
+        `metadata` parameter can be sent as a stringified JSON or in pure JSON.
+        """
+        if not isinstance(metadata, dict):
+            try:
+                metadata = json.loads(metadata)
+            except TypeError:
+                # Let the validator returns an explicit message to user that
+                # `metadata` is required
+                pass
+            except ValueError:
+                raise serializers.ValidationError({
+                    'metadata': _('JSON is invalid')
+                })
+
+        return metadata
+
+    def _validate_base64Encoded(self, attr: dict, metadata: dict):  # noqa
+        base64_encoded = attr['base64Encoded']
+        metadata = self._validate_metadata(metadata)
+
+        try:
+            media_content = base64_encoded[base64_encoded.index('base64') + 7:]
+        except ValueError:
+            raise serializers.ValidationError({
+                'base64Encoded': _('Invalid content')
+            })
+
+        attr.update({
+            'content': ContentFile(base64.decodebytes(media_content.encode()),
+                                   name=metadata['filename'])
+        })
+
+    def _validate_content(self, attr: dict, metadata: dict):
+        try:
+            attr['content']
+        except KeyError:
+            raise serializers.ValidationError({
+                'content': _('No files have been submitted')
+            })
+
+        metadata['filename'] = attr['content'].name
+
+    def _validate_duplicate(self, filename: str, field_name: str):
+
+        if self.__file_type == AssetFile.FORM_MEDIA:
+            view = self.context.get('view')
+            asset = getattr(view, 'asset', self.context.get('asset'))
+            # File name must be unique
+            if AssetFile.objects.filter(asset=asset,
+                                        date_deleted__isnull=True,
+                                        metadata__filename=filename).exists():
```

Comment:
1. the kpi database cannot enforce uniqueness because the filename comes from either the `content` (for an uploaded file) or the `metadata` (for a URL)
1. however, kobocat does enforce uniqueness (of uploaded files and URLs separately?)
1. there are possibly sticky situations where a user would have to go into the kobocat ui and remove a file to solve a conflict(?)
1. can we generally fix this by adding some new parameter to the kobocat endpoint for form media file uploads that would allow overwriting a file with the same name instead of failing with an `IntegrityError` (or whatever it does now)

Reply:
What is your our final decision for this?

Reply:
let's deal with it when we handle #2675. i just copied my comment over there.

:warning: github says that #2675 **will be closed** by this PR, but this work doesn't yet do anything to fetch old media from KoBoCAT, right?